Keep track of countries mapped per quiz

CallaJune opened the pull request:
## What

This change is less of a full change and more of a needed change partially implemented so we can discuss how we'd like to carry it out. I added keep track of the countries mapped in one single game and include a variable in the backend which stores the countries mapped and their frequency.

## How

* Added a `country` column to the locations table in the database.
* Created and maintained a state variable called `gameStats`, a dictionary (so we can add more game-level stats in the future) which holds another dictionary called `countryCounts`.
* Throughout the game, I use `countryCounts` as a hash map storing the countries mapped and the number of times they are mapped. I then pass this variable into the results page.

## Tests

I visualized the result of the hash map (stringified) in the results page in order to verify that it worked, adding in fake country values. 

<img width="494" alt="Screen Shot 2021-01-01 at 11 03 56 AM" src="https://user-images.githubusercontent.com/6993618/103442239-690d2400-4c22-11eb-8553-825ee16a0278.png">

## Discussion

Before I get into how my skeletal implementation may relate to choices we should make, I want to note that my hash map `countryCounts` is not updated in the most efficient manner. I take the value of the state, add a new entry, and then set the value of the state to the updated dictionary. I understand that this is not efficient. But it seems that `setState` has to be used in order to change the value of the state. Do you have any alternative methods?

### Country Counts

I have updated `locations` to include a column (of type text) called `country`. It may be good practice to hold a country code instead of using the country name. If so, do you have a specific code format that you usually use? I was looking at [this one](https://unece.org/cefact/unlocode-code-list-country-and-territory). If we did this, we could have a structure of some sort which mapped codes to country names. The `locations` table would include the country code, and we could then get the country name from that. This would make the "keys" used neater and more memory efficient (though not significantly so) as some country names are long, could be formatted in different ways, and have special characters. Later on, we may even want to include translations of the site, in which case we may translate the names of some countries, and using codes would make that simpler. 

However, I didn't want to implement this yet as what is easiest depends on how data is coming into this system. What are all of the data flows into the `locations` table, and what steps would need to be taken to add a country name to them? In some, would a country name be provided which would have to be converted into a country code? I wanted to learn about this more before implementing something which could potentially make this more complicated.

The `countryCounts` dictionary is like so:

```
{
    country: count
}
```
If `gameStats.country_counts` is passed into the result page, we can then access countries' counts using `props.countryCounts.Ecuador` or `props.countryCounts.EC` if we use [these country codes](https://unece.org/cefact/unlocode-code-list-country-and-territory). This dictionary was designed in a way that would make it easy for us to create some sort of map highlighting countries that were mapped during that game. We could alternatively present the user with their top n countries, which would require a relatively trivial change. It all depends on how we design the result page, and I believe this gives us some flexibility.

### Game Stats

I added a state variable, a dictionary that assumes we will want to keep track of game-specific data. We have not yet settled on what the result page will look like, but I presume we will not want to _only_ include data about the user's impact overall. I designed this to give us the most flexibility. 

###

lacabra commented:
Thanks @CallaJune. Everything what you propose sounds very good. To your question about country codes, I strongly suggest using [ISO 3166 Alpha-3](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-3) codes (you were suggesting Alpha-2). I prefer Alpha-3 because I find 3-character codes are more intuitive to identify most of the time than 2-character codes, and while the intent is that humans will not look directly at these codes, it makes debugging and testing easier in some cases.

CallaJune commented:
@lacabra I have added a `countries` table mapping country names to country codes. 

Country codes were generated using [this json file](https://github.com/lukes/ISO-3166-Countries-with-Regional-Codes/blob/master/slim-2/slim-2.json), though I changed "alpha-1" to "code". 

I've removed showing country names to each quiz question, but here's a screenshot from when I added `{props.question.country_name}` to the `quiz` component. 

<img width="496" alt="Screen Shot 2021-01-22 at 8 58 52 PM" src="https://user-images.githubusercontent.com/6993618/105566114-76d13900-5cf8-11eb-8745-2b4c903da520.png">

lacabra reviewed (changes requested):
Thanks @CallaJune, I would ask the following:

- include a copy of the [original json](https://github.com/lukes/ISO-3166-Countries-with-Regional-Codes/blob/master/slim-2/slim-2.json) in the scripts folder in our repo
- do not modify "alpha-2" to "code", instead use the original file as is, and modify the code to refer to that field as `alpha-2`
- document the origin of this file in our repo, for example add another bullet point to the Architecture section.

All this is so that if in the future we need to update the file (i.e. we notice that the file changes at origin, it will be more straightforward to do so, without duplicating all the small changes we do on our end). I argue this eases maintainability of this code moving forward.

CallaJune commented:
@lacabra Agreed. I made this change and changed the source to what I see as a more reliable source long-term, Data Hub vs. someone's personal Github account.

lacabra commented on db/dbOps.js:
```diff
@@ -58,22 +61,36 @@ async function createTables() {
+	await pool.query(`
+		CREATE TABLE countries(
+			country_code TEXT,
+			country_name TEXT
```
could you add here an additional condition:
`unique (country_code)` 
which would help validate the data upon loading the data from the json. It's an extra precaution.

Disregard this comment if we drop this table altogether, so address the previous comment first, to decide whether this one is still relevant.

lacabra replied:
This comment is still valid. If you disagree, kindly ague against it. I want to make sure there is no duplication in the `country_code`. If there is, we will know that there is something wrong with the JSON file.

lacabra commented on README.md:
````diff
@@ -37,6 +37,8 @@ PGPASSWORD="POSTGRES_PASSWORD"
 PGDATABASE="POSTGRES_DATABASE"
 ```
 
+You must populate the database by using the the script `db/dbOps.js`, which you can run from the root directory with the command `node db/dbOps.js`. In order to load the table `countries`, download [the JSON file of country names and ISO-3166 alpha-2 codes here](https://datahub.io/core/country-list) from Data Hub and uncomment the `loadTables();` command at the bottom of `db/dbOps.js`. 
````
Please add the additional information on where this file should be saved to: `scripts/countryCodes.json`

lacabra commented on db/dbOps.js:
```diff
@@ -58,22 +61,36 @@ async function createTables() {
 			ip INET
 		);
 	`);
+	await pool.query(`
```
Given that the pairing of country codes and country names is "static" once we have downloaded the JSON and will not change, it is not entirely clear to me what's the benefit of storing it on the database. You can import it anywhere in the code where you need the pairing by importing the json file. The database only needs to store the `country_code` in table `locations`. 

Can you make the case for why should we store it in the DB?

CallaJune replied:
Good point. I think we should absolutely place it in a JSON file and import it. I'll look into doing that in this architecture.

lacabra commented on db/dbOps.js:
```diff
@@ -58,22 +61,36 @@ async function createTables() {
 		console.log(res)
 	}
+
+	for (let i = 0; i < countryCodes.length; i++) {
```
Same for whether this table is needed altogether, see previous comment on the table creation.

lacabra commented on pages/api/getLocations/[uuid].js:
```diff
@@ -13,13 +13,16 @@ export default async (req, res) => {
 							locations.lat,
-							locations.lon
+							locations.lon,
+							countries.country_name,
```
Following the argument above, you don't need to retrieve `country_name`; you can pull it from the code directly when it needs to be displayed to the user.

lacabra commented on pages/mapping.js:
```diff
@@ -12,10 +12,11 @@ export default function mapping() {
 	const [answerCount, setAnswerCount] = useState(0);
 	const [counter, setCounter] = useState(0);
-	const [question, setQuestion] = useState({ id: 0, lat: 0, lon: 0, answer: ''});
+	const [question, setQuestion] = useState({ id: 0, lat: 0, lon: 0, country_name: '', country_code: '', answer: ''});
```
In the state, you only need to keep one of the two, that would be `country_code`. Make the case otherwise.

CallaJune commented:
@lacabra I have changed this to pull the country code from a JSON file. It appears on the quiz when I add this to the quiz page: `{countryName}`

I pulled the JSON from the original source, which included a list of JSON objects. I have included reformatted version of this and documentation and a script in case it needs to be regenerated.

<img width="509" alt="Screen Shot 2021-01-26 at 9 05 35 PM" src="https://user-images.githubusercontent.com/6993618/105932110-5cf66580-601a-11eb-9399-17b71f55e9ad.png">

lacabra commented on README.md:
```diff
@@ -37,6 +37,18 @@ PGPASSWORD="POSTGRES_PASSWORD"
+The [countries.json](data/countries.json) is included in this repository to associate country codes with country names. If this file needs to be updated or replaced, follow these instructions:
+
+1. Download [the JSON file of country names and ISO-3166 alpha-2 codes here](https://datahub.io/core/country-list) from Data Hub. Place the JSON file in `scripts/` of this repository and call it `countryCodes.json`.
+2. From the root directory, run `node createCountryCodesJson.js` to run the script that creates the JSON file the game requires and places it in `data/`. 
```
Very minor comment, but it seems this line should say:

From the root directory, run `node scripts/createCountryCodesJson.js` 

(note the missing `scripts/` text)

lacabra reviewed (changes requested):
Thanks @CallaJune, everything looks much better now. There are still two minor comments I would like you to address 🙏

CallaJune commented:
It appears I simply forgot to make the change in dbOps. My intention was to simply get rid of the countries table altogether. If we are using the countries json file for the country names, would that make sense to you — since we no longer have any use for it in the table? The way I create the countries json file, it is at one point a JavaScript dictionary and so that takes care of duplicate country codes. 

> On Jan 26, 2021, at 11:25 PM, Victor <notifications@github.com> wrote:
> 
> ﻿
> @lacabra requested changes on this pull request.
> 
> Thanks @CallaJune, everything looks much better now. There are still two minor comments I would like you to address 🙏
> 
> In README.md:
> 
> > @@ -37,6 +37,18 @@ PGPASSWORD="POSTGRES_PASSWORD"
>  PGDATABASE="POSTGRES_DATABASE"
>  ```
>  
> +The [countries.json](data/countries.json) is included in this repository to associate country codes with country names. If this file needs to be updated or replaced, follow these instructions:
> +
> +1. Download [the JSON file of country names and ISO-3166 alpha-2 codes here](https://datahub.io/core/country-list) from Data Hub. Place the JSON file in `scripts/` of this repository and call it `countryCodes.json`.
> +2. From the root directory, run `node createCountryCodesJson.js` to run the script that creates the JSON file the game requires and places it in `data/`. 
> Very minor comment, but it seems this line should say:
> 
> From the root directory, run node scripts/createCountryCodesJson.js
> 
> (note the missing scripts/ text)
> 
> In db/dbOps.js:
> 
> > @@ -58,22 +61,36 @@ async function createTables() {
>  			ip INET
>  		);
>  	`);
> +	await pool.query(`
> +		CREATE TABLE countries(
> +			country_code TEXT,
> +			country_name TEXT
> This comment is still valid. If you disagree, kindly ague against it. I want to make sure there is no duplication in the country_code. If there is, we will know that there is something wrong with the JSON file.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

lacabra commented:
@CallaJune You are absolutely right. Remove the table creation altogether, and we are in agreement. Thanks! 👍

lacabra reviewed (approved):
Looks terrific! 🙏 Thank you for addressing all my comments; much appreciated.
